File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-java/src/main/java/com/sdkwork/Mail/metadata/mailProviderSelection.java
package com.sdkwork.Mail.metadata;

import java.util.List;

public record MailProviderSelection(String providerKey, MailProviderSelectionSource source) {

  public enum MailProviderSelectionSource {
    provider_url,
    provider_key,
    tenant_override,
    deployment_profile,
    default_provider
  }

  public record ParsedMailProviderUrl(String providerKey, String rawUrl) {
  }

  public record MailProviderSelectionRequest(
      String providerUrl,
      String providerKey,
      String tenantOverrideProviderKey,
      String deploymentProfileProviderKey
  ) {
  }

  public static final List<String> mail_PROVIDER_SELECTION_SOURCES = List.of(
      "provider_url",
      "provider_key",
      "tenant_override",
      "deployment_profile",
      "default_provider"
  );

  public static final List<String> mail_PROVIDER_SELECTION_PRECEDENCE = List.of(
      "provider_url",
      "provider_key",
      "tenant_override",
      "deployment_profile",
      "default_provider"
  );

  public static ParsedMailProviderUrl parseMailProviderUrl(String providerUrl) {
    var trimmed = providerUrl.trim();
    if (!trimmed.startsWith("Mail:") || !trimmed.contains("://")) {
      throw new IllegalArgumentException("Invalid Mail provider URL: " + providerUrl);
    }

    return new ParsedMailProviderUrl(
        trimmed.substring(4, trimmed.indexOf("://")).toLowerCase(),
        providerUrl
    );
  }

  public static MailProviderSelection resolveMailProviderSelection(
      MailProviderSelectionRequest request
  ) {
    return resolveMailProviderSelection(request, MailProviderCatalog.DEFAULT_mail_PROVIDER_KEY);
  }

  public static MailProviderSelection resolveMailProviderSelection(
      MailProviderSelectionRequest request,
      String defaultProviderKey
  ) {
    var resolvedRequest = request == null
        ? new MailProviderSelectionRequest(null, null, null, null)
        : request;

    if (hasText(resolvedRequest.providerUrl())) {
      return new MailProviderSelection(
          parseMailProviderUrl(resolvedRequest.providerUrl()).providerKey(),
          MailProviderSelectionSource.provider_url
      );
    }

    if (hasText(resolvedRequest.providerKey())) {
      return new MailProviderSelection(
          resolvedRequest.providerKey().trim(),
          MailProviderSelectionSource.provider_key
      );
    }

    if (hasText(resolvedRequest.tenantOverrideProviderKey())) {
      return new MailProviderSelection(
          resolvedRequest.tenantOverrideProviderKey().trim(),
          MailProviderSelectionSource.tenant_override
      );
    }

    if (hasText(resolvedRequest.deploymentProfileProviderKey())) {
      return new MailProviderSelection(
          resolvedRequest.deploymentProfileProviderKey().trim(),
          MailProviderSelectionSource.deployment_profile
      );
    }

    return new MailProviderSelection(
        defaultProviderKey,
        MailProviderSelectionSource.default_provider
    );
  }

  private static boolean hasText(String value) {
    return value != null && !value.trim().isEmpty();
  }
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-kotlin/src/main/kotlin/com/sdkwork/Mail/metadata/mailProviderSelection.kt
package com.sdkwork.Mail.metadata

enum class MailProviderSelectionSource {
    provider_url,
    provider_key,
    tenant_override,
    deployment_profile,
    default_provider,
}

data class ParsedMailProviderUrl(
    val providerKey: String,
    val rawUrl: String,
)

data class MailProviderSelection(
    val providerKey: String,
    val source: MailProviderSelectionSource,
)

data class MailProviderSelectionRequest(
    val providerUrl: String? = null,
    val providerKey: String? = null,
    val tenantOverrideProviderKey: String? = null,
    val deploymentProfileProviderKey: String? = null,
)

val mail_PROVIDER_SELECTION_SOURCES: List<String> = listOf(
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
)

val mail_PROVIDER_SELECTION_PRECEDENCE: List<String> = listOf(
    "provider_url",
    "provider_key",
    "tenant_override",
    "deployment_profile",
    "default_provider",
)

private fun hasMailProviderSelectionText(value: String?): Boolean = value != null && value.isNotBlank()

fun parseMailProviderUrl(providerUrl: String): ParsedMailProviderUrl {
    val trimmed = providerUrl.trim()
    require(trimmed.startsWith("Mail:") && trimmed.contains("://")) {
        "Invalid Mail provider URL: $providerUrl"
    }

    return ParsedMailProviderUrl(
        providerKey = trimmed.substring(4, trimmed.indexOf("://")).lowercase(),
        rawUrl = providerUrl,
    )
}

fun resolveMailProviderSelection(
    request: MailProviderSelectionRequest = MailProviderSelectionRequest(),
    defaultProviderKey: String = MailProviderCatalog.DEFAULT_mail_PROVIDER_KEY,
): MailProviderSelection {
    if (hasMailProviderSelectionText(request.providerUrl)) {
        return MailProviderSelection(
            providerKey = parseMailProviderUrl(request.providerUrl!!).providerKey,
            source = MailProviderSelectionSource.provider_url,
        )
    }

    if (hasMailProviderSelectionText(request.providerKey)) {
        return MailProviderSelection(
            providerKey = request.providerKey!!.trim(),
            source = MailProviderSelectionSource.provider_key,
        )
    }

    if (hasMailProviderSelectionText(request.tenantOverrideProviderKey)) {
        return MailProviderSelection(
            providerKey = request.tenantOverrideProviderKey!!.trim(),
            source = MailProviderSelectionSource.tenant_override,
        )
    }

    if (hasMailProviderSelectionText(request.deploymentProfileProviderKey)) {
        return MailProviderSelection(
            providerKey = request.deploymentProfileProviderKey!!.trim(),
            source = MailProviderSelectionSource.deployment_profile,
        )
    }

    return MailProviderSelection(
        providerKey = defaultProviderKey,
        source = MailProviderSelectionSource.default_provider,
    )
}
